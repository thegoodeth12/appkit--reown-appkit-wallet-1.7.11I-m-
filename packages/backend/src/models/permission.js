import Base from '@/models/base.js';
import permissionCatalog from '@/helpers/permission-catalog.ee.js';

class Permission extends Base {
  static tableName = 'permissions';

  static jsonSchema = {
    type: 'object',
    required: ['roleId', 'action', 'subject'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      roleId: { type: 'string', format: 'uuid' },
      action: { type: 'string', minLength: 1 },
      subject: { type: 'string', minLength: 1 },
      conditions: { type: 'array', items: { type: 'string' } },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static filter(permissions) {
    const sanitizedPermissions = permissions.filter((permission) => {
      const { action, subject, conditions } = permission;

      const relevantAction = this.findAction(action);
      const validSubject = this.isSubjectValid(subject, relevantAction);
      const validConditions = this.areConditionsValid(conditions);

      return relevantAction && validSubject && validConditions;
    });

    return sanitizedPermissions;
  }

  static findAction(action) {
    return permissionCatalog.actions.find(
      (actionCatalogItem) => actionCatalogItem.key === action
    );
  }

  static isSubjectValid(subject, action) {
    return action && action.subjects.includes(subject);
  }

  static areConditionsValid(conditions) {
    return conditions.every((condition) => this.isConditionValid(condition));
  }

  static isConditionValid(condition) {
    return !!permissionCatalog.conditions.find(
      (conditionCatalogItem) => conditionCatalogItem.key === condition
    );
  }
}

export default Permission;
