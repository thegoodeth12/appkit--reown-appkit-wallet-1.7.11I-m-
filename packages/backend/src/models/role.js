import { ValidationError } from 'objection';
import Base from '@/models/base.js';
import Permission from '@/models/permission.js';
import User from '@/models/user.js';
import SamlAuthProvider from '@/models/saml-auth-provider.ee.js';
import NotAuthorizedError from '@/errors/not-authorized.js';

class Role extends Base {
  static tableName = 'roles';

  static jsonSchema = {
    type: 'object',
    required: ['name'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      name: { type: 'string', minLength: 1 },
      description: { type: ['string', 'null'], maxLength: 255 },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    users: {
      relation: Base.HasManyRelation,
      modelClass: User,
      join: {
        from: 'roles.id',
        to: 'users.role_id',
      },
    },
    permissions: {
      relation: Base.HasManyRelation,
      modelClass: Permission,
      join: {
        from: 'roles.id',
        to: 'permissions.role_id',
      },
    },
  });

  static get virtualAttributes() {
    return ['isAdmin'];
  }

  get isAdmin() {
    return this.name === 'Admin';
  }

  static async findAdmin() {
    return await this.query().findOne({ name: 'Admin' });
  }

  async preventAlteringAdmin() {
    const currentRole = await Role.query().findById(this.id);

    if (currentRole.isAdmin) {
      throw new NotAuthorizedError('The admin role cannot be altered!');
    }
  }

  async deletePermissions() {
    return await this.$relatedQuery('permissions').delete();
  }

  async createPermissions(permissions) {
    if (permissions?.length) {
      const validPermissions = Permission.filter(permissions).map(
        (permission) => ({
          ...permission,
          roleId: this.id,
        })
      );

      await Permission.query().insert(validPermissions);
    }
  }

  async updatePermissions(permissions) {
    await this.deletePermissions();

    await this.createPermissions(permissions);
  }

  async updateWithPermissions(data) {
    const { name, description, permissions } = data;

    await this.updatePermissions(permissions);

    await this.$query().patchAndFetch({
      id: this.id,
      name,
      description,
    });

    return await this.$query()
      .leftJoinRelated({
        permissions: true,
      })
      .withGraphFetched({
        permissions: true,
      });
  }

  async deleteWithPermissions() {
    await this.deletePermissions();

    return await this.$query().delete();
  }

  async assertNoRoleUserExists() {
    const userCount = await this.$relatedQuery('users').limit(1).resultSize();
    const hasUsers = userCount > 0;

    if (hasUsers) {
      throw new ValidationError({
        data: {
          role: [
            {
              message: `All users must be migrated away from the "${this.name}" role.`,
            },
          ],
        },
        type: 'ValidationError',
      });
    }
  }

  async assertNoConfigurationUsage() {
    const samlAuthProviderUsingDefaultRole = await SamlAuthProvider.query()
      .where({
        default_role_id: this.id,
      })
      .limit(1)
      .first();

    if (samlAuthProviderUsingDefaultRole) {
      throw new ValidationError({
        data: {
          samlAuthProvider: [
            {
              message:
                'You need to change the default role in the SAML configuration before deleting this role.',
            },
          ],
        },
        type: 'ValidationError',
      });
    }
  }

  async assertRoleIsNotUsed() {
    await this.assertNoRoleUserExists();

    await this.assertNoConfigurationUsage();
  }

  async $beforeUpdate(opt, queryContext) {
    await super.$beforeUpdate(opt, queryContext);

    await this.preventAlteringAdmin();
  }

  async $beforeDelete(queryContext) {
    await super.$beforeDelete(queryContext);

    await this.preventAlteringAdmin();

    await this.assertRoleIsNotUsed();
  }
}

export default Role;
