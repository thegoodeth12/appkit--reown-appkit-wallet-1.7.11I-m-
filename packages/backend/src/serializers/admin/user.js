import userSerializer from '@/serializers/user.js';

const adminUserSerializer = (user) => {
  const userData = userSerializer(user);

  userData.acceptInvitationUrl = user.acceptInvitationUrl;

  return userData;
};

export default adminUserSerializer;
