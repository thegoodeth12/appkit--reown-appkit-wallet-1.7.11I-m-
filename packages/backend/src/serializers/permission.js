const permissionSerializer = (permission) => {
  return {
    id: permission.id,
    roleId: permission.roleId,
    action: permission.action,
    subject: permission.subject,
    conditions: permission.conditions,
    createdAt: permission.createdAt.getTime(),
    updatedAt: permission.updatedAt.getTime(),
  };
};

export default permissionSerializer;
