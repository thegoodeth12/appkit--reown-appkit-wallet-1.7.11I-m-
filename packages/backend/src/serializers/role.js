import permissionSerializer from '@/serializers/permission.js';

const roleSerializer = (role) => {
  let roleData = {
    id: role.id,
    name: role.name,
    key: role.key,
    description: role.description,
    createdAt: role.createdAt.getTime(),
    updatedAt: role.updatedAt.getTime(),
    isAdmin: role.isAdmin,
  };

  if (role.permissions?.length > 0) {
    roleData.permissions = role.permissions.map((permission) =>
      permissionSerializer(permission)
    );
  }

  return roleData;
};

export default roleSerializer;
