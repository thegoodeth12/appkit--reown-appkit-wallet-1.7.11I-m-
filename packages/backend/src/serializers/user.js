import roleSerializer from '@/serializers/role.js';
import permissionSerializer from '@/serializers/permission.js';
import appConfig from '@/config/app.js';

const userSerializer = (user) => {
  let userData = {
    id: user.id,
    email: user.email,
    createdAt: user.createdAt.getTime(),
    updatedAt: user.updatedAt.getTime(),
    status: user.status,
    fullName: user.fullName,
  };

  if (user.role) {
    userData.role = roleSerializer(user.role);
  }

  if (user.permissions?.length > 0) {
    userData.permissions = user.permissions.map((permission) =>
      permissionSerializer(permission)
    );
  }

  if (appConfig.isCloud && user.trialExpiryDate) {
    userData.trialExpiryDate = user.trialExpiryDate;
  }

  return userData;
};

export default userSerializer;
