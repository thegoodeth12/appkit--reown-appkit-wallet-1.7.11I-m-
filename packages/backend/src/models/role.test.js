import { describe, it, expect, vi } from 'vitest';
import Role from '@/models/role.js';
import Base from '@/models/base.js';
import Permission from '@/models/permission.js';
import User from '@/models/user.js';
import { createRole } from '@/factories/role.js';
import { createPermission } from '@/factories/permission.js';
import { createUser } from '@/factories/user.js';
import { createSamlAuthProvider } from '@/factories/saml-auth-provider.ee.js';

describe('Role model', () => {
  it('tableName should return correct name', () => {
    expect(Role.tableName).toBe('roles');
  });

  it('jsonSchema should have correct validations', () => {
    expect(Role.jsonSchema).toMatchSnapshot();
  });

  it('relationMappingsshould return correct associations', () => {
    const relationMappings = Role.relationMappings();

    const expectedRelations = {
      users: {
        relation: Base.HasManyRelation,
        modelClass: User,
        join: {
          from: 'roles.id',
          to: 'users.role_id',
        },
      },
      permissions: {
        relation: Base.HasManyRelation,
        modelClass: Permission,
        join: {
          from: 'roles.id',
          to: 'permissions.role_id',
        },
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  it('virtualAttributes should return correct attributes', () => {
    expect(Role.virtualAttributes).toStrictEqual(['isAdmin']);
  });

  describe('isAdmin', () => {
    it('should return true for admin named role', () => {
      const role = new Role();
      role.name = 'Admin';

      expect(role.isAdmin).toBe(true);
    });

    it('should return false for not admin named roles', () => {
      const role = new Role();
      role.name = 'User';

      expect(role.isAdmin).toBe(false);
    });
  });

  it('findAdmin should return admin role', async () => {
    const createdAdminRole = await createRole({ name: 'Admin' });

    const adminRole = await Role.findAdmin();

    expect(createdAdminRole).toStrictEqual(adminRole);
  });

  describe('preventAlteringAdmin', () => {
    it('preventAlteringAdmin should throw an error when altering admin role', async () => {
      const role = await createRole({ name: 'Admin' });

      await expect(() => role.preventAlteringAdmin()).rejects.toThrowError(
        'The admin role cannot be altered!'
      );
    });

    it('preventAlteringAdmin should not throw an error when altering non-admin roles', async () => {
      const role = await createRole({ name: 'User' });

      expect(await role.preventAlteringAdmin()).toBe(undefined);
    });
  });

  it("deletePermissions should delete role's permissions", async () => {
    const role = await createRole({ name: 'User' });
    await createPermission({ roleId: role.id });

    await role.deletePermissions();

    expect(await role.$relatedQuery('permissions')).toStrictEqual([]);
  });

  describe('createPermissions', () => {
    it('should create permissions', async () => {
      const role = await createRole({ name: 'User' });

      await role.createPermissions([
        { action: 'read', subject: 'Flow', conditions: [] },
      ]);

      expect(await role.$relatedQuery('permissions')).toMatchObject([
        {
          action: 'read',
          subject: 'Flow',
          conditions: [],
        },
      ]);
    });

    it('should call Permission.filter', async () => {
      const role = await createRole({ name: 'User' });

      const permissions = [{ action: 'read', subject: 'Flow', conditions: [] }];

      const permissionFilterSpy = vi
        .spyOn(Permission, 'filter')
        .mockReturnValue(permissions);

      await role.createPermissions(permissions);

      expect(permissionFilterSpy).toHaveBeenCalledWith(permissions);
    });
  });

  it('updatePermissions should delete existing permissions and create new permissions', async () => {
    const permissionsData = [
      { action: 'read', subject: 'Flow', conditions: [] },
    ];

    const deletePermissionsSpy = vi
      .spyOn(Role.prototype, 'deletePermissions')
      .mockResolvedValueOnce();
    const createPermissionsSpy = vi
      .spyOn(Role.prototype, 'createPermissions')
      .mockResolvedValueOnce();

    const role = await createRole({ name: 'User' });

    await role.updatePermissions(permissionsData);

    expect(deletePermissionsSpy.mock.invocationCallOrder[0]).toBeLessThan(
      createPermissionsSpy.mock.invocationCallOrder[0]
    );

    expect(deletePermissionsSpy).toHaveBeenNthCalledWith(1);
    expect(createPermissionsSpy).toHaveBeenNthCalledWith(1, permissionsData);
  });

  describe('updateWithPermissions', () => {
    it('should update role along with given permissions', async () => {
      const role = await createRole({ name: 'User' });
      await createPermission({
        roleId: role.id,
        subject: 'Flow',
        action: 'read',
        conditions: [],
      });

      const newRoleData = {
        name: 'Updated user',
        description: 'Updated description',
        permissions: [
          {
            action: 'manage',
            subject: 'Flow',
            conditions: [],
          },
        ],
      };

      await role.updateWithPermissions(newRoleData);

      const roleWithPermissions = await role
        .$query()
        .leftJoinRelated({ permissions: true })
        .withGraphFetched({ permissions: true });

      expect(roleWithPermissions).toMatchObject(newRoleData);
    });
  });

  describe('deleteWithPermissions', () => {
    it('should delete role along with given permissions', async () => {
      const role = await createRole({ name: 'User' });
      await createPermission({
        roleId: role.id,
        subject: 'Flow',
        action: 'read',
        conditions: [],
      });

      await role.deleteWithPermissions();

      const refetchedRole = await role.$query();
      const rolePermissions = await Permission.query().where({
        roleId: role.id,
      });

      expect(refetchedRole).toBe(undefined);
      expect(rolePermissions).toStrictEqual([]);
    });
  });

  describe('assertNoRoleUserExists', () => {
    it('should reject with an error when the role has users', async () => {
      const role = await createRole({ name: 'User' });
      await createUser({ roleId: role.id });

      await expect(() => role.assertNoRoleUserExists()).rejects.toThrowError(
        `All users must be migrated away from the "User" role.`
      );
    });

    it('should resolve when the role does not have any users', async () => {
      const role = await createRole();

      expect(await role.assertNoRoleUserExists()).toBe(undefined);
    });
  });

  describe('assertNoConfigurationUsage', () => {
    it('should reject with an error when the role is used in configuration', async () => {
      const role = await createRole();
      await createSamlAuthProvider({ defaultRoleId: role.id });

      await expect(() =>
        role.assertNoConfigurationUsage()
      ).rejects.toThrowError(
        'samlAuthProvider: You need to change the default role in the SAML configuration before deleting this role.'
      );
    });

    it('should resolve when the role does not have any users', async () => {
      const role = await createRole();

      expect(await role.assertNoConfigurationUsage()).toBe(undefined);
    });
  });

  it('assertRoleIsNotUsed should call assertNoRoleUserExists and assertNoConfigurationUsage', async () => {
    const role = new Role();

    const assertNoRoleUserExistsSpy = vi
      .spyOn(role, 'assertNoRoleUserExists')
      .mockResolvedValue();

    const assertNoConfigurationUsageSpy = vi
      .spyOn(role, 'assertNoConfigurationUsage')
      .mockResolvedValue();

    await role.assertRoleIsNotUsed();

    expect(assertNoRoleUserExistsSpy).toHaveBeenCalledOnce();
    expect(assertNoConfigurationUsageSpy).toHaveBeenCalledOnce();
  });

  describe('$beforeDelete', () => {
    it('should call preventAlteringAdmin', async () => {
      const role = await createRole({ name: 'User' });

      const preventAlteringAdminSpy = vi
        .spyOn(role, 'preventAlteringAdmin')
        .mockResolvedValue();

      await role.$query().delete();

      expect(preventAlteringAdminSpy).toHaveBeenCalledOnce();
    });

    it('should call assertRoleIsNotUsed', async () => {
      const role = await createRole({ name: 'User' });

      const assertRoleIsNotUsedSpy = vi
        .spyOn(role, 'assertRoleIsNotUsed')
        .mockResolvedValue();

      await role.$query().delete();

      expect(assertRoleIsNotUsedSpy).toHaveBeenCalledOnce();
    });
  });
});
