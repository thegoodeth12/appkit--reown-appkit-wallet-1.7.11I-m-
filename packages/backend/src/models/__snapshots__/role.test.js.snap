// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Role model > jsonSchema should have correct validations 1`] = `
{
  "properties": {
    "createdAt": {
      "type": "string",
    },
    "description": {
      "maxLength": 255,
      "type": [
        "string",
        "null",
      ],
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "name": {
      "minLength": 1,
      "type": "string",
    },
    "updatedAt": {
      "type": "string",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;
